### DISCUSSION TITLE
Add support for JPEG XL

### DISCUSSION THREAD

**Discussion by gunta** [libvips/libvips#2055] [Ideas]
Why:
JPEG XL is the future of all image formats.

Explanation here:
https://www.slideshare.net/cloudinarymarketing/imagecon-2019-jon-sneyer

Reference implementation here:
https://gitlab.com/wg1/jpeg-xl

**Comment by jcupitt** [comment]
Sure, looks interesting. PRs etc. very welcome.

**Comment by kmartinez** [comment]
nice to see AVX2 being used! probably requires a delicate config/make as so many CPUs don't have it but as you say - it could be the future!

**Comment by dkam** [comment]
Integrating JPEG-XL is beyond me, but [ImageMagick](https://github.com/ImageMagick/ImageMagick/blob/main/coders/jxl.c) has integrated it.

**Comment by fredprodibi** [comment]
The image stream is now frozen,

is it a good time to re-eval the integration of JPEG-XL?

**Comment by tomByrer** [comment]
For [web browsers, AVIF](https://caniuse.com/avif) has much better support.  Also unsure if there is hardware support directly for AVIF, but the underlining AV1 does have HW decoding in Apple's M1 So "JPEG XL is the future of all image formats." is an overstatement.  Without the web, JXL won't become popular (like JPEG2000), even if [it is great](https://cloudinary.com/blog/time_for_next_gen_codecs_to_dethrone_jpeg?utm_source=reddit&utm_medium=social&utm_campaign=time-for-next-gen-codecs-to-dethrone-jpeg).

I would hope it is supported, but keep the above in mind for priority.

**Reply by jcupitt** [reply]
libvips has pretty good AVIF support already, so I don't see that as a blocker to JXL. I plan to have a look at this before 8.11, though  of course I don't know how much work it'll be.

**Reply by 190n** [reply]
> the underlining AV1 does have HW decoding in Apple's M1

The M1 does not support hardware AV1 decoding, but it performs decently in software. RDNA2, Ampere, Tiger Lake, Rocket Lake, and some ARM SoCs (although none from Apple or Qualcomm) do support hardware AV1 decoding.

**Comment by renchap** [comment]
FYI Chrome intends to ship JPEG XL support (behind a flag for now): https://groups.google.com/a/chromium.org/g/blink-dev/c/WjCKcBw219k/m/BK9qXyz1BwAJ

They mention a strong interest from multiple groups on this.

**Comment by jcupitt** [comment]
There's a branch here with some work:

https://github.com/libvips/libvips/tree/add-jxl

It compiles, but doesn't run yet.

jpeg-xl encoding is *not* quick! For a 1450x2048 pixel jpg image I see:

```
$ vips colourspace ~/pics/k2.jpg x.pfm[strip] scrgb
$ time ./encode_oneshot x.pfm ~/pics/k2.jxl 
real	0m3.128s
user	0m30.568s
sys	0m0.243s
```

30s of CPU time to encode (!!).

**Reply by jcupitt** [reply]
Ah that was a debug build, it's much quicker in release:

```
$ time ./encode_oneshot x.pfm x.jxl 
real	0m0.456s
user	0m1.041s
sys	0m0.391s
```

About 10x slower than libjpeg-turbo.

```
$ time vips copy x.ppm x.jpg
real	0m0.047s
user	0m0.089s
sys	0m0.009s
```

**Comment by jcupitt** [comment]
OK, all done, there's a PR here with working load and save:

https://github.com/libvips/libvips/pull/2181

Though it has a range of issues and libjxl seems a little immature still. I doubt if anyone will want to use this in production.

**Comment by renchap** [comment]
Here is an interesting comment about JPEG XL from Facebook: https://bugzilla.mozilla.org/show_bug.cgi?id=1539075#c18

TL;DR: encoding with a quality of 6 is as fast as JPEG, decoding with multithreading is as fast as JPEG, it has more features / same efficiency as other modern image codecs based on video (like AVIF).

It looks very promising for a web usage!

**Reply by jcupitt** [reply]
Yes, my tests line up with that -- very fast (compared to avif anyway), and great compression.

It does use a lot of memory, on the downside, but improvements are supposed to be coming. I would guess it'll be another year or so of work to get to production.

**Reply by flawrist** [reply]
A lot of memory for compression only (like FLAC) or decompression requires the same amount??

**Comment by flawrist** [comment]
I am learning about options of `vips jxlsave`. For example, `--distance` is more or less clear thanks to [that manual][1], i.e. its default to 1 means near-lossless output, but what is some decode speed `--tier`?

[1]: https://gitlab.com/wg1/jpeg-xl/-/raw/main/doc/man/cjxl.txt

**Reply by jcupitt** [reply]
My understanding is that the tiers set the types of decompression that the compressor will target. JXL has several decompression systems, some slow and some fast, and compressors can use a mix of them to try to recreate the image.

**Comment by richardmonette** [comment]
Does libvips support progressive JXL encoding? i.e. the 

```
-p, --progressive
    Enable progressive/responsive decoding
```

flag from cjxl?

**Reply by jcupitt** [reply]
No, not yet.

**Reply by richardmonette** [reply]
Am I correct in understanding that libvips supports JXL through the libjxl library? If that is the case, is it a matter of adding the argument to libvips and then passing it through as a parameter to libjxl?

**Reply by jcupitt** [reply]
libvips only exposes a subset of the libjxl encoding options. You can see the API we support here:

https://www.libvips.org/API/current/VipsForeignSave.html#vips-jxlsave

**Comment by t4k** [comment]
My goal will be to create pyramid TIFFs with JPEG-XL compression. I don’t see a compression option for JPEG-XL just yet in VipsForeignTiffCompression. Is that on the radar?

**Reply by jcupitt** [reply]
I'd be a little cautious -- JXL is still not stable enough for production use (IMO) and implementing JXL in TIFF would mean our TIFF loader could no longer be trusted with internet data.

A PR for this would be great, but it would have to be off by default for now.

**Reply by AngelaDMerkel** [reply]
I have just tried this on the latest master and see that it still doesn't work. This would be welcomed. I have had trouble with moving metadata from TIFF to JXL no matter what I have tried, and moving from TIFF container to TIFF container would likely solve this for me while offering most of the space savings associated with lossless JXL.

**Comment by astrolox** [comment]
This became relevant to me recently, so I thought I'd share here some interesting links for anyone else who ends up reading this discussion thread.

Chromium had an experimental implementation of JPEG XL behind a flag, but that it was removed;
https://groups.google.com/a/chromium.org/g/blink-dev/c/WjCKcBw219k/m/xX-NnWtTBQAJ

Which was commented on by Free Software Foundation;
https://www.fsf.org/blogs/community/googles-decision-to-deprecate-jpeg-xl-emphasizes-the-need-for-browser-choice-and-free-formats

Although there is still hope that they may reimplement it in the future;
https://bugs.chromium.org/p/chromium/issues/detail?id=1416530
https://bugs.chromium.org/p/chromium/issues/detail?id=1178058

**Reply by t4k** [reply]
Yes, those developments have been extremely disappointing to me. Thanks for the links.

**Reply by septatrix** [reply]
JPEG XL support in chrome has since been reinstated

**Comment by renchap** [comment]
Apple announced that Safari will now support JPEG XL on all OS, this is great news!

**Reply by dmitrizzle** [reply]
Nice! I've been waiting for lossy 16-bit per channel support, looks like it's coming soon.

**Comment by fredprodibi** [comment]
May be it's time to include jpgxl to the web build? it would help adoption.

**Comment by flawrist** [comment]
Dear VIPS developers, I'm puzzled about how lossless thing is done here:

* [vips.exe](https://github.com/libvips/build-win64-mxe/releases/tag/v8.15.0-rc2) copy [image.jpg](https://s7.krakenfiles.com/uploads/09-11-2023/OSMbbh2YVA/image.jpg) image.vips.jxl --lossless
works for a long time, 4244645 vs 7994675 bytes (+88%)
* [cjxl.exe](https://github.com/libjxl/libjxl/releases/tag/v0.8.2) image.jpg image.cjxl.jxl
works almost instantly, 4244645 vs 3466260 bytes (-18%)
  ```console
  $ cjxl image.jpg image.cjxl.jxl
  JPEG XL encoder v0.8.2 954b460 [SSSE3,Unknown]
  Note: Implicit-default for JPEG is lossless-transcoding.
  Read JPEG image with 4244645 bytes.
  Encoding [Container | JPEG, lossless transcode, effort: 7 | JPEG reconstruction data],
  Compressed to 3466260 bytes including container

  $ djxl image.cjxl.jxl image.djxl.jpg
  JPEG XL decoder v0.8.2 954b460 [SSSE3,Unknown]
  Read 3466260 compressed bytes.
  Reconstructed to JPEG.
  4048 x 3036, 5.89 MP/s [5.89, 5.89], 2.03 MB/s [2.03, 2.03], 1 reps, 2 threads.

  $ xxhsum *.jpg
  c8336de1b739ed3e  image.jpg
  c8336de1b739ed3e  image.djxl.jpg
  
  $ jxlinfo image.vips.jxl
  JPEG XL image, 4048x3036, (possibly) lossless, 8-bit RGB
  Color space: 596-byte ICC profile, CMM type: "", color space: "RGB ", rendering intent: 0
  
  $ jxlinfo image.cjxl.jxl
  JPEG XL file format container (ISO/IEC 18181-2)                                          
  JPEG XL image, 4048x3036, (possibly) lossless, 8-bit RGB                                 
  Color space: 596-byte ICC profile, CMM type: "", color space: "RGB ", rendering intent: 0
  JPEG bitstream reconstruction data available                                             
  Brotli-compressed Exif metadata: 27514 compressed bytes
  ```

Why?

**Reply by jcupitt** [reply]
Hello @sergeevabc,

We wrote the libvips jxl interface code back in the libjxl-0.4 days (I think?), and the libjxl API has evolved since then. We have some work on this bit of libvips scheduled for libvips 8.16, see eg.:

https://github.com/libvips/libvips/pull/3712

**Reply by DarthSim** [reply]
@sergeevabc You're mixing up lossless transcoding and lossless encoding.

`cjxl image.jpg image.cjxl.jxl` does lossless transcoding from JPEG to JPEG XL. This allows you to create a JPEG XL image from JPEG data without decoding it. Since the data is not decoded and encoded back, there are no additional losses. And since `cjxl` doesn't need to decode/encode the image, it converts it to JPEG XL momentarily.

`vips copy image.jpg image.vips.jxl --lossless` does lossless encoding. Vips works with raw pixels, so it needs to decode an image before doing stuff with it. So, it can't losslessly transcode JPEG to JPEG XL. At least, it would require a lot of work and a huge amount of duct tape to make it do so. Decoding/encoding of an image takes some time, especially lossless encoding enabled with the `--lossless` flag. And of course, lossless encoding produces huge files.

**Comment by jpc-ae** [comment]
Are there any plans to move jxl to the web/default binaries soon? Or is there hesitancy to use libjxl in favour of waiting for a stable rust implimentation?

**Comment by yoshiyoshyosh** [comment]
I'm writing this here to have it kept track somewhere: The way vips currently loads jpeg xl images doesn't seem to take as good advantage of its parallelizability as djxl.

For an example, I made a 16kx16k lossy jxl with pyvips via 
```
Image.new_from_memory(random.randbytes(16384*16384*3),16384,16384,3,'uchar')
.gaussblur(5)
.write_to_file('/tmp/vips/16k.jxl',distance=1,effort=4)
```

and compared runtimes:

```
$ time djxl 16k.jxl --disable_output
Decoded to pixels.
16384 x 16384, 203.819 MP/s, 12 threads.
real  0m1.678s
user  0m8.754s
sys   0m3.505s

$ time ./main 16k.jxl
...
vips_foreign_load_jxl_process: seen JXL_DEC_NEED_IMAGE_OUT_BUFFER
vips_foreign_load_jxl_process: starting ...
wall time: 2.706470                                          // <-- The slow part!
vips_foreign_load_jxl_process: seen JXL_DEC_FULL_IMAGE

width: 16384 | height: 16384 | size: 805306368 | overall wall time: 2.981394
real  0m3.298s
user  0m8.620s
sys   0m3.134s
```
`main` here is just a program that does a `vips_image_new_from_file()` into `write_to_memory()`. I inserted some wall time prints for each stage of jxl loading--before and after the while loop in `vips_foreign_load_jxl_process`.

`djxl` spikes all my threads by ~50%, while `vips` doesn't break ~20% for each worker. It has something to do with the `INPUT_BUFFER`'s existence and/or size, as upping it to `32768` instead of `4096` decreases the wall time for `JXL_DEC_FULL_IMAGE` to ~1.8 seconds. I'm not entirely sure why this input buffer exists in the first place though? My first thought was streaming inputs for low memory consumption, but there's an upfront `malloc` of the uncompressed pixel data size before it even gets to frame decoding anyway, so again, not sure. Some more insight would be greatly appreciated.

I can maybe poke at it once I grok more of the loading pipeline, but this was already a distraction of a distraction for me, so maybe later...

**Reply by jcupitt** [reply]
Hi @yoshiyoshyosh, I've not looked at this code for a while, but from memory the jxl decoder is data driven, not demand driven, so it's a poor fit for libvips. We are forced to decode the whole image to memory in one shot and without threading.

Most image load libraries have an API something like:

```C
int jpeg_get_next_scanline(Compressioninfo *info, uchar *scanline);
```

Which decodes the next scanline from the compression info into the memory supplied by the caller, with callbacks in `info` to `read()` bytes from the input stream as required. When libvips sees a request for another chunk of pixels, it can just call this a few times to decompress the next part of the image it needs.

libjxl is completely different, it has a state machine running on the input stream. As bytes arrive, they are added to the compression engine, and that emits events when things like frame-is-complete happen. Image decompression is driven by the arrival of data, not by demand for pixels.

libvips currently takes the easy way out and just decompresses the whole image to ram, then copies pixels from that to the output image on demand.

We could instead run the libjxl decompression in a background thread and have a set of semaphores to stop and start that thread as image pixels are requested. But that would be a lot of work to build and test, and no one's found the energy.

I think there was talk of a new libjxl read API which might be a better fit, but I've not looked into it recently.

**Reply by jcupitt** [reply]
(libjxl write works like other image format libraries, so libvips can do streaming jxl save)

**Comment by jpc-ae** [comment]
I take it [the threaded runner](https://libjxl.readthedocs.io/en/latest/api_threads.html#_CPPv423JxlThreadParallelRunnerPvPv18JxlParallelRunInit22JxlParallelRunFunction8uint32_t8uint32_t) does not help? Would switching to jxl-rs for read operations help?

**Reply by jcupitt** [reply]
Hi @jpc-ae, yes we use the threaded runner:

https://github.com/libvips/libvips/blob/master/libvips/foreign/jxlload.c#L201-L208

It's funny it doesn't help more. Maybe using `JxlThreadParallelRunnerDefaultNumWorkerThreads()` would be better.

Does jxl-rs have a "pull" API? I've not looked. Something to decode a specific tile would be ideal (eg. like libtiff).

**Reply by yoshiyoshyosh** [reply]
The threaded runner isn't helping much (especially on larger images) because the loader requests 4096 pixels at a time. JXL can only really be effectively parallelized in its DC/AC blocks (VarDCT, 2048x2048/256x256) or its LF groups (Modular, 128x128 to 1024x1024). Basically the sections outlined in the table of contents. I'm not exactly sure what order this request goes in, but it's certainly bottlenecking the decoding process by having the actual pushing of memory go through a single thread it looks like? Again, haven't 100% groked the loader architecture yet.

Using `JxlThreadParallelRunnerDefaultNumWorkerThreads()` wouldn't change anything. It's just a wrapper around `std::thread::hardware_concurrency()`. It returns the same amount of threads as `vips_concurrency_get()` for me.

I poked a bit around libjxl and it does appear there's no real per-pixel or per-group pull API... you'd have to make one manually as of right now. Not ideal!

**Reply by jpc-ae** [reply]
My understanding is this is a result of how its progressive decoding mode works. For images saved this way (preferable for the web), the entire image is streamed with greater and greater detail refinement meaning tile data would be split across the file in such a way that finding and combining the parts would be non-trivial. Please correct me if I'm wrong.

**Reply by jpc-ae** [reply]
I do wonder how switching to `JxlResizableParallelRunner` would affect things. Unfortunately, I'm not really set up to test this at the moment. Just throwing ideas.

**Reply by jcupitt** [reply]
> The threaded runner isn't helping much (especially on larger images) because the loader requests 4096 pixels at a time. JXL can only really be effectively parallelized in its DC/AC blocks (VarDCT, 2048x2048/256x256) or its LF groups (Modular, 128x128 to 1024x1024).

libvips loads frames with this loop:

https://github.com/libvips/libvips/blob/master/libvips/foreign/jxlload.c#L550-L592

So it throws bytes from the file into the hopper until it sees a `JXL_DEC_FULL_IMAGE` event. Is there some way to ask for smaller or lower-res image regions to be reconstructed? 

I should probably read the current `djxl` sources!

**Comment by hlindset** [comment]
Hi! I had a quick look at  `jxlsave.c`, and it seems to write chunks to `VipsTarget` as they are released, and seeks using `vips_target_seek()`. The `set_finalized_position` callback is a no-op. This seems to make `jxlsave_target` a little bit awkward for non-seekable targets, such as pipes and http streams, since bytes emitted may be changed later.

Would it be feasible to have `jxlsave.c` support non-seekable targets by staging bytes internally, only flushing bytes up to the finalized position as it advances?
